client/state/upgrade.go

```diff
@@ -0,0 +1,299 @@
+package state
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/boltdb/bolt"
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/helper/boltdd"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/ugorji/go/codec"
+)
+
+// NeedsUpgrade returns true if the BoltDB needs upgrading or false if it is
+// already up to date.
+func NeedsUpgrade(bdb *bolt.DB) (bool, error) {
+	needsUpgrade := true
+	err := bdb.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(metaBucketName)
+		if b == nil {
+			// No meta bucket; upgrade
+			return nil
+		}
+
+		v := b.Get(metaVersionKey)
+		if len(v) == 0 {
+			// No version; upgrade
+			return nil
+		}
+
+		if !bytes.Equal(v, metaVersion) {
+			// Version exists but does not match. Abort.
+			return fmt.Errorf("incompatible state version. expected %q but found %q",
+				string(metaVersion), string(v))
+		}
+
+		// Version matches! Assume migrated!
+		needsUpgrade = false
+		return nil
+	})
+
+	return needsUpgrade, err
+}
+
+// addMeta adds version metadata to BoltDB to mark it as upgraded and
+// should be run at the end of the upgrade transaction.
+func addMeta(tx *bolt.Tx) error {
+	// Create the meta bucket if it doesn't exist
+	bkt, err := tx.CreateBucketIfNotExists(metaBucketName)
+	if err != nil {
+		return err
+	}
+
+	return bkt.Put(metaVersionKey, metaVersion)
+}
+
+// backupDB backs up the existing state database prior to upgrade overwriting
+// previous backups.
+func backupDB(bdb *bolt.DB, dst string) error {
+	fd, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+
+	return bdb.View(func(tx *bolt.Tx) error {
+		if _, err := tx.WriteTo(fd); err != nil {
+			fd.Close()
+			return err
+		}
+
+		return fd.Close()
+	})
+}
+
+// UpgradeSchema upgrades the boltdb schema. Example 0.8 schema:
+//
+//	* allocations
+//	  * 15d83e8a-74a2-b4da-3f17-ed5c12895ea8
+//	    * echo
+//	       - simple-all (342 bytes)
+//	     - alloc (2827 bytes)
+//	     - alloc-dir (166 bytes)
+//	     - immutable (15 bytes)
+//	     - mutable (1294 bytes)
+//
+func UpgradeAllocs(logger hclog.Logger, tx *boltdd.Tx) error {
+	btx := tx.BoltTx()
+	allocationsBucket := btx.Bucket(allocationsBucketName)
+	if allocationsBucket == nil {
+		// No state!
+		return nil
+	}
+
+	// Gather alloc buckets and remove unexpected key/value pairs
+	allocBuckets := [][]byte{}
+	cur := allocationsBucket.Cursor()
+	for k, v := cur.First(); k != nil; k, v = cur.Next() {
+		if v != nil {
+			logger.Warn("deleting unexpected key in state db",
+				"key", string(k), "value_bytes", len(v),
+			)
+
+			if err := cur.Delete(); err != nil {
+				return err
+			}
+			continue
+		}
+
+		allocBuckets = append(allocBuckets, k)
+	}
+
+	for _, allocBucket := range allocBuckets {
+		allocID := string(allocBucket)
+
+		bkt := allocationsBucket.Bucket(allocBucket)
```

Comment:
Could we parse out the buckets in the first loop? This would allow us to fail faster in the case of really weird state

Reply:
I need to keep the key (`allocID` above) around for logging/errors. Sadly bolt doesn't have an API for retrieving the name from a `*tx.Bucket`.